cog_fixtures/cog.py

```diff
@@ -0,0 +1,118 @@
+"""cog_fixtures.cog module"""
+import abc
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, Optional, Sequence, Union
+
+import affine
+import rasterio
+from fake_geo_images.fakegeoimages import FakeGeoImage
+from rasterio.crs import CRS
+from rasterio.io import MemoryFile
+from rio_cogeo.cogeo import cog_translate
+
+
+@dataclass  # type: ignore
+class ImageBase(abc.ABC):
+    """base image class"""
+
+    def __enter__(self):
+        """support context management"""
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """support context management"""
+        self.close()
+
+    @abc.abstractmethod
+    def __post_init__(self):
+        """post init hook"""
+        ...
+
+    @abc.abstractmethod
+    def close(self):
+        """close resources"""
+        ...
+
+
+@dataclass
+class FakeImage(ImageBase):
+    """fake rasterio image"""
+
+    driver: str
+    width: int
+    height: int
+    count: int
+    dtype: str
+    crs: CRS
+    transform: affine.Affine
+    nodata: Union[float, int] = -1.0
+
+    def __post_init__(self):
+        """post init hook"""
+        fpath, _ = FakeGeoImage(
+            xsize=self.width,
+            ysize=self.height,
+            num_bands=self.count,
+            data_type=self.dtype,
+            out_dir=Path("/vsimem"),
+            crs=self.crs.to_epsg(),
+            nodata=self.nodata,
+        ).create(transform=self.transform)
+        self.handle = rasterio.open(fpath)
+
+    def close(self):
+        """close resources"""
+        self.handle.close()
+
+
+@dataclass
+class FakeCog(ImageBase):
+    """fake cloud optimized geotiff"""
+
+    src_img: FakeImage
+    dst_kwargs: Dict
+    indexes: Optional[Sequence[int]] = None
+    nodata: Optional[Union[str, int, float]] = None
```

Comment:
nodata has to be `Union[int, float]`. `str` is when a user pass `nan` in the CLI which we translate to `numpy.nan` which is a `float`